### ISSUE TITLE
Spurious errors in out var error scenarios

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#12058]
The `out var` implementation depends on an internal method, `BuildArgumentsForErrorRecovery`, to do work for semantic analysis that it wasn't designed to do. This is fragile and results in incorrect behavior. An example test case is as follows:

```cs
class Program
{
    public static void Main(string[] args)
    {
        if (M(s: out var s))
        {
            string s2 = s;
        }
    }
    public static bool M(int i, out string s)
    {
        s = i.ToString();
        return true;
    }
}
```

This should produce the single error

```none
(5,13): error CS7036: There is no argument given that corresponds to the required formal parameter 'i' of 'Program.M(int, out string)'
        if (M(s: out var s))
```

but the implementation also produces the spurious error

```none
(7,25): error CS0029: Cannot implicitly convert type 'int' to 'string'
            string s2 = s;
```

Here is this issue in the form of a unit test:

```cs
        [Fact]
        public void MissingArgumentAndNamedOutVarArgument()
        {
            var source =
@"class Program
{
    public static void Main(string[] args)
    {
        if (M(s: out var s))
        {
            string s2 = s;
        }
    }
    public static bool M(int i, out string s)
    {
        s = i.ToString();
        return true;
    }
}
";
            var compilation = CreateCompilationWithMscorlib(source,
                                                            options: TestOptions.ReleaseExe,
                                                            parseOptions: TestOptions.Regular.WithOutVarFeature());
            compilation.VerifyDiagnostics(
                // (5,13): error CS7036: There is no argument given that corresponds to the required formal parameter 'i' of 'Program.M(int, out string)'
                //         if (M(s: out var s))
                Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "M").WithArguments("i", "Program.M(int, out string)").WithLocation(5, 13)
                );
        }
```

**Comment by AlekseyTs** [comment]
The compiler behaves as intended.

**Comment by gafter** [comment]
Binding to an out variable of type `string` should not result in a variable of type `int`. We need to adjust our intentions.

**Comment by AlekseyTs** [comment]
> Binding to an out variable of type string should not result in a variable of type int.

That is true, but in my opinion the code being compiled is not targeting out variable of type string.

We can either fail type inference for the out var in BuildArgumentsForErrorRecovery or infer the type from the parameter of the only candidate that we have. I've chosen to infer the type, we can fail the inference instead. Each approach has its pluses and minuses. In any case, I think BuildArgumentsForErrorRecovery is the right place to finalize the inference, whether it is a success or a failure.

**Comment by gafter** [comment]
> That is true, but in my opinion the code being compiled is not targeting out variable of type string.

If it is targeting the parameter of type int, why is the compiler complaining that an argument for that parameter is missing? That simply doesn't make sense.

**Comment by AlekseyTs** [comment]
That is a good point, I missed the fact that name arguments are involved. Then I would say, there is a general preexisting problem in BuildArgumentsForErrorRecovery - it doesn't take named nature of arguments into account. It looks like lambdas are affected as well. The heuristic of matching arguments to parameters in that function should be improved.

**Comment by gafter** [comment]
@AlekseyTs Agreed, there is room for improvement in `BuildArgumentsForErrorRecovery`. It wasn't designed to be part of the behavior of semantic analysis, but now it is being used for that.

**Comment by AlekseyTs** [comment]
> It wasn't designed to be part of the behavior of semantic analysis

@gafter I think what that function is doing with lambdas doesn't agree with this statement. 
You clearly demonstrated that there is an existing problem with how it matches arguments to parameters, which already affects lambdas. BTW, are you going to fix that, since you are fixing error recovery issues around lambdas? 
Clearly, whoever introduced the function didn't do it specifically for Out Variable Declarations. But I see nothing wrong in adjusting its purpose in life to do that. Whether the function should fail inference or not for Out Variable Declaration is debatable. The fact that there is something wrong in doing things related to Out Local Declarations in BuildArgumentsForErrorRecovery is yet to be demonstrated. 

BTW, thank you for testing the feature!

**Comment by gafter** [comment]
`BuildArgumentsForErrorRecovery` doesn't affect the compiler's semantic analysis today.

**Comment by AlekseyTs** [comment]
> BuildArgumentsForErrorRecovery doesn't affect the compiler's semantic analysis today.

Of course it does, it affects SemanticModel behavior in lambdas.

**Comment by gafter** [comment]
I said "semantic analysis" in order to distinguish it from "semantic model API". The compiler's analysis (i.e. the code that handles diagnostics and code generation) doesn't look at the result of `BuildArgumentsForErrorRecovery`.

**Comment by AlekseyTs** [comment]
I believe that the only thing that this function affects with respect to Out Variable Declarations is additional diagnostics reported **in an error recovery** situation. All this is dialable. So far you haven't demonstrated any design flaw.

**Comment by AlekseyTs** [comment]
> BuildArgumentsForErrorRecovery ... mutating some bound nodes in the process (second problem)

@gafter Do you have **any factual information** to support this claim? 
I am fairly confident that BuildArgumentsForErrorRecovery, or any other code within compiler doesn't mutate any bound nodes specific to Out Variable Declarations feature after the nodes were created.

**Comment by gafter** [comment]
> > BuildArgumentsForErrorRecovery ... mutating some bound nodes in the process (second problem)

> @gafter Do you have **any factual information** to support this claim? 
I am fairly confident that BuildArgumentsForErrorRecovery, or any other code within compiler doesn't mutate any bound nodes specific to Out Variable Declarations feature after the nodes were created.

That may be so. I assumed that was the case, since similar code for binding `Deconstruct` in the `features/tuples` branch does mutate the bound nodes, and the code was modeled after the out var implementation.

**Comment by AlekseyTs** [comment]
>  I assumed that was the case, since similar code for binding Deconstruct in the features/tuples branch does mutate the bound nodes, and the code was modeled after the out var implementation.

That is a "No" then. Thanks for confirming.

**Comment by AlekseyTs** [comment]
The issue has been addressed already, a regression unit-test has been added.
